src/core/Akka/Dispatch/MessageQueues/BoundedMessageQueue.cs

```diff
@@ -7,12 +7,21 @@ namespace Akka.Dispatch.MessageQueues
     /// <summary>An Bounded mailbox message queue.</summary>
     public class BoundedMessageQueue : MessageQueue, BoundedMessageQueueSemantics
     {
-        //TODO: Implement BoundedMessageQueue. Currently it's just a copy of UnboundedMessageQueue
-        private readonly ConcurrentQueue<Envelope> _queue = new ConcurrentQueue<Envelope>();
+        private readonly BlockingCollection<Envelope> _queue;
+
+        public BoundedMessageQueue()
+        {
+            _queue = new BlockingCollection<Envelope>();
+        }
+
+        public BoundedMessageQueue(int boundedCapacity)
+        {
+            _queue = new BlockingCollection<Envelope>(boundedCapacity);
+        }
 
         public void Enqueue(Envelope envelope)
         {
-            _queue.Enqueue(envelope);
+            _queue.Add(envelope);
```

Comment:
Use this overload instead:
https://msdn.microsoft.com/en-us/library/dd287149(v=vs.110).aspx

`TryAdd(Envelope,Timespan)`

And use the `PushTimeout` property.
If I recall correctly, if the timeout occurs, the message should be sent to deadletters, not sure, check Scala impl/docs.